cpp/velox/shuffle/VeloxSortBasedShuffleWriter.cc

```diff
@@ -0,0 +1,349 @@
+    RETURN_NOT_OK(initFromRowVector(*rv.get()));
+    RETURN_NOT_OK(doSort(rv, partitionWriter_.get()->options().sortBufferMaxSize));
+  } else {
+    auto veloxColumnBatch = VeloxColumnarBatch::from(veloxPool_.get(), cb);
+    VELOX_CHECK_NOT_NULL(veloxColumnBatch);
+    facebook::velox::RowVectorPtr rv;
+    START_TIMING(cpuWallTimingList_[CpuWallTimingFlattenRV]);
+    rv = veloxColumnBatch->getFlattenedRowVector();
+    ;
+    END_TIMING();
+    if (partitioner_->hasPid()) {
+      auto pidArr = getFirstColumn(*rv);
+      START_TIMING(cpuWallTimingList_[CpuWallTimingCompute]);
+      RETURN_NOT_OK(partitioner_->compute(pidArr, rv->size(), batches_.size(), rowVectorIndexMap_));
+      END_TIMING();
+      auto strippedRv = getStrippedRowVector(*rv);
+      RETURN_NOT_OK(initFromRowVector(*strippedRv));
+      RETURN_NOT_OK(doSort(strippedRv, partitionWriter_.get()->options().sortBufferMaxSize));
+    } else {
+      RETURN_NOT_OK(initFromRowVector(*rv));
+      START_TIMING(cpuWallTimingList_[CpuWallTimingCompute]);
+      RETURN_NOT_OK(partitioner_->compute(nullptr, rv->size(), batches_.size(), rowVectorIndexMap_));
+      END_TIMING();
+      RETURN_NOT_OK(doSort(rv, partitionWriter_.get()->options().sortBufferMaxSize));
+    }
+  }
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortBasedShuffleWriter::evictBatch(
+    uint32_t partitionId,
+    std::ostringstream* output,
+    facebook::velox::OStreamOutputStream* out,
+    facebook::velox::RowTypePtr* rowTypePtr) {
+  int64_t rawSize = batch_->size();
+  batch_->flush(out);
+  const std::string& outputStr = output->str();
+  RETURN_NOT_OK(partitionWriter_->evict(partitionId, rawSize, outputStr.c_str(), outputStr.size()));
+  batch_.reset();
+  output->clear();
+  output->str("");
+  batch_ = std::make_unique<facebook::velox::VectorStreamGroup>(veloxPool_.get());
+  batch_->createStreamTree(*rowTypePtr, options_.bufferSize, &serdeOptions_);
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortBasedShuffleWriter::evictRowVector(uint32_t partitionId) {
+  int32_t rowNum = 0;
+  const int32_t maxBatchNum = options_.bufferSize;
+  auto rowTypePtr = std::static_pointer_cast<const facebook::velox::RowType>(rowType_.value());
+  std::ostringstream output;
+  facebook::velox::OStreamOutputStream out(&output);
+
+  if (options_.partitioning != Partitioning::kSingle) {
+    if (auto it = rowVectorIndexMap_.find(partitionId); it != rowVectorIndexMap_.end()) {
+      auto rowVectorIndex = it->second;
+      const int32_t outputSize = rowVectorIndex.size();
+
+      std::unordered_map<int32_t, std::vector<facebook::velox::IndexRange>> groupedIndices;
+
+      for (int start = 0; start < outputSize; start++) {
+        const int32_t vectorIndex = static_cast<int32_t>(rowVectorIndex.at(start) >> 32);
+        const int32_t rowIndex = static_cast<int32_t>(rowVectorIndex.at(start) & 0xFFFFFFFFLL);
+        groupedIndices[vectorIndex].push_back({rowIndex, 1});
```

Comment:
The mapping from the input vector index + row index is already ordered. Can we directly build the index range in Partitioner::compute? Then the adjacent row index of the same partition id can be merged.

Reply:
@marin-ma  ```Partitioner::compute``` belongs to the core part of the cpp, it is not appropriate to introduce Velox-related ```IndexRange```, I will implement this part of the logic in ```evictRowVector```.